packages/remix-server-runtime/__tests__/responses-test.ts

```diff
@@ -1,4 +1,230 @@
+      !done;
+      { done, value } = await reader.read(), chunkCount++
+    ) {
+      if (chunkCount > 3) {
+        throw new Error("Too many chunks seen in deferred body");
+      }
+
+      decodedChunks.push(decoder.decode(value));
+    }
+
+    expect(decodedChunks.length).toBe(3);
+
+    let [initialChunkType, initialData] = chunkTypeAndData(decodedChunks[0]);
+    expect(initialChunkType).toBe("initial");
+    expect(initialData).toEqual({
+      foo: "remix",
+      bar: DEFERRED_PROMISE_PREFIX + "bar",
+      baz: DEFERRED_PROMISE_PREFIX + "baz",
+    });
+
+    let [secondChunkType, secondChunkData] = chunkTypeAndData(decodedChunks[1]);
+    expect(secondChunkType).toBe("data");
+    let secondChunkDataKeys = Object.keys(secondChunkData);
+    expect(secondChunkDataKeys.length).toBe(1);
+    expect(secondChunkData[secondChunkDataKeys[0]]).toEqual(
+      await data[secondChunkDataKeys[0]]
+    );
+
+    let [thirdChunkType, thirdChunkData] = chunkTypeAndData(decodedChunks[2]);
+    expect(thirdChunkType).toBe("data");
+    let thirdChunkDataKeys = Object.keys(thirdChunkData);
+    expect(thirdChunkDataKeys.length).toBe(1);
+    expect(thirdChunkData[thirdChunkDataKeys[0]]).toEqual(
+      await data[thirdChunkDataKeys[0]]
+    );
+  });
+
+  it("encodes the response body with object with rejected promises", async () => {
+    let data = {
+      foo: "remix",
+      bar: Promise.reject(new Error("rejected")),
+      baz: Promise.resolve({ sub: "value" }),
+    };
+    let response = deferred(data);
+    expect(await response.json()).toEqual({
+      foo: "remix",
+      bar: DEFERRED_PROMISE_PREFIX + "bar",
+      baz: DEFERRED_PROMISE_PREFIX + "baz",
+    });
+    expect(response.deferred.bar).toBeDefined();
+    expect(response.deferred.baz).toBeDefined();
+    await expect(response.deferred.bar).rejects.toThrowError("rejected");
+    expect(await response.deferred.baz).toEqual({ sub: "value" });
+
+    expect(response.body).toBeDefined();
+    let reader = response.body.getReader();
+    let decoder = new TextDecoder();
+    let decodedChunks: string[] = [];
+    for (
+      let { done, value } = await reader.read(), chunkCount = 0;
+      !done;
+      { done, value } = await reader.read(), chunkCount++
+    ) {
+      if (chunkCount > 3) {
+        throw new Error("Too many chunks seen in deferred body");
+      }
+
+      decodedChunks.push(decoder.decode(value));
+    }
+
+    expect(decodedChunks.length).toBe(3);
+
+    let [initialChunkType, initialData] = chunkTypeAndData(decodedChunks[0]);
+    expect(initialChunkType).toBe("initial");
+    expect(initialData).toEqual({
+      foo: "remix",
+      bar: DEFERRED_PROMISE_PREFIX + "bar",
+      baz: DEFERRED_PROMISE_PREFIX + "baz",
+    });
+
+    let [secondChunkType, secondChunkData] = chunkTypeAndData(decodedChunks[1]);
+    let secondChunkDataKeys = Object.keys(secondChunkData);
+    expect(secondChunkDataKeys.length).toBe(1);
+    if (secondChunkType === "error") {
```

Comment:
Maybe store all chunks in an array, then go find which one is the error/data to avoid duplicating if/else if/else statements.

Reply:
These test have been cleaned up and moved to the shared package.